File: go/mcp_helper_golden_vectors_test.go
// Cross-language MCP bridge helper-parity fixture test
// (`MCP_BRIDGE_SDK_PLAN.md` P1-P3 conformance).
//
// Loads `net/crates/net/tests/cross_lang_mcp/helper_vectors.json` — the
// canonical fixture the Rust source-of-truth verifier
// (`adapters/mcp/tests/helper_golden_vectors.rs`) validates — and asserts
// the Go net.Classify / net.LowerTool wrappers produce the same results. A
// failure here means the Go binding drifted from the core.

package net

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"testing"
)

func mcpFixturePath(t *testing.T) string {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	dir := filepath.Dir(thisFile)
	// The published go/ module is at the repo root; fixtures are at
	// ../net/crates/net/tests/cross_lang_mcp/.
	return filepath.Join(dir, "..", "net", "crates", "net", "tests", "cross_lang_mcp", "helper_vectors.json")
}

type mcpClassifyCase struct {
	Name               string            `json:"name"`
	Program            string            `json:"program"`
	Args               []string          `json:"args"`
	Envs               map[string]string `json:"envs"`
	CredentialOverride *string           `json:"credential_override"`
	Force              bool              `json:"force"`
	ExpectedStatus     string            `json:"expected_status"`
}

type mcpLowerCase struct {
	Name             string          `json:"name"`
	Tool             json.RawMessage `json:"tool"`
	ServerVersion    string          `json:"server_version"`
	CredentialStatus string          `json:"credential_status"`
	Substitutability string          `json:"substitutability"`
	Expected         json.RawMessage `json:"expected"`
}

type mcpHelperFixture struct {
	Classify []mcpClassifyCase `json:"classify"`
	Lower    []mcpLowerCase    `json:"lower"`
}

func loadMcpHelperFixture(t *testing.T) mcpHelperFixture {
	t.Helper()
	raw, err := os.ReadFile(mcpFixturePath(t))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var f mcpHelperFixture
	if err := json.Unmarshal(raw, &f); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	return f
}

func TestClassifyGoldenVectors(t *testing.T) {
	for _, c := range loadMcpHelperFixture(t).Classify {
		override := ""
		if c.CredentialOverride != nil {
			override = *c.CredentialOverride
		}
		got, err := Classify(c.Program, c.Args, c.Envs, override, c.Force)
		if err != nil {
			t.Fatalf("[%s] Classify error: %v", c.Name, err)
		}
		if got != c.ExpectedStatus {
			t.Errorf("[%s] classify = %q, want %q", c.Name, got, c.ExpectedStatus)
		}
	}
}

// normalizeToInterface reshapes a lowered DTO into the fixture's comparison
// shape — the descriptor's input_schema / output_schema JSON strings become
// parsed *_object values — then round-trips through JSON to interface{} so
// the comparison is by value (all numbers become float64 on both sides).
func normalizeLowered(t *testing.T, lowered LoweredTool) interface{} {
	t.Helper()
	var desc map[string]interface{}
	if err := json.Unmarshal(lowered.Descriptor, &desc); err != nil {
		t.Fatalf("descriptor not JSON: %v", err)
	}
	parseSchema := func(key string) interface{} {
		if s, ok := desc[key].(string); ok && s != "" {
			var obj interface{}
			if err := json.Unmarshal([]byte(s), &obj); err != nil {
				t.Fatalf("%s not JSON: %v", key, err)
			}
			return obj
		}
		return nil
	}
	desc["input_schema_object"] = parseSchema("input_schema")
	desc["output_schema_object"] = parseSchema("output_schema")
	delete(desc, "input_schema")
	delete(desc, "output_schema")

	got := map[string]interface{}{
		"tool_id":         lowered.ToolID,
		"mcp_name":        lowered.McpName,
		"bridge_metadata": lowered.BridgeMetadata,
		"descriptor":      desc,
	}
	blob, err := json.Marshal(got)
	if err != nil {
		t.Fatalf("re-marshal got: %v", err)
	}
	var norm interface{}
	if err := json.Unmarshal(blob, &norm); err != nil {
		t.Fatalf("normalize got: %v", err)
	}
	return norm
}

func TestLowerGoldenVectors(t *testing.T) {
	for _, c := range loadMcpHelperFixture(t).Lower {
		lowered, err := LowerTool(string(c.Tool), c.ServerVersion, c.CredentialStatus, c.Substitutability)
		if err != nil {
			t.Fatalf("[%s] LowerTool error: %v", c.Name, err)
		}
		got := normalizeLowered(t, lowered)

		var want interface{}
		if err := json.Unmarshal(c.Expected, &want); err != nil {
			t.Fatalf("[%s] parse expected: %v", c.Name, err)
		}
		if !reflect.DeepEqual(got, want) {
			gotJSON, _ := json.MarshalIndent(got, "", "  ")
			wantJSON, _ := json.MarshalIndent(want, "", "  ")
			t.Errorf("[%s] lower DTO mismatch\n got: %s\nwant: %s", c.Name, gotJSON, wantJSON)
		}
	}
}

File: net/crates/net/bindings/node/test/mcp_helper_golden_vectors.test.ts
// Cross-language MCP bridge helper-parity fixture test
// (`MCP_BRIDGE_SDK_PLAN.md` P1-P3 conformance).
//
// Loads `crates/net/tests/cross_lang_mcp/helper_vectors.json` — the
// canonical fixture the Rust source-of-truth verifier
// (`adapters/mcp/tests/helper_golden_vectors.rs`) validates — and asserts
// the Node classifyMcpServer / lowerMcpTool bindings produce the same
// results. A failure here means the Node binding drifted from the core.

import { readFileSync } from 'node:fs'
import { join } from 'node:path'

import { describe, expect, it } from 'vitest'

import { classifyMcpServer, lowerMcpTool } from '../index'

const FIXTURE_PATH = join(
  __dirname,
  '..',
  '..',
  '..',
  'tests',
  'cross_lang_mcp',
  'helper_vectors.json',
)

interface ClassifyCase {
  name: string
  program: string
  args: string[]
  envs: Record<string, string>
  credential_override: string | null
  force: boolean
  expected_status: string
}

interface LowerCase {
  name: string
  tool: unknown
  server_version: string
  credential_status: string
  substitutability: string
  expected: unknown
}

const FIXTURE = JSON.parse(readFileSync(FIXTURE_PATH, 'utf8')) as {
  classify: ClassifyCase[]
  lower: LowerCase[]
}

// Reshape a lowered DTO into the fixture's comparison shape: descriptor is a
// JSON string here (the napi convention), and inside it input_schema /
// output_schema are JSON strings — parse them all so the comparison is by
// value. Field names map from camelCase (napi) to the fixture's snake_case.
function normalize(result: {
  toolId: string
  mcpName: string
  descriptor: string
  bridgeMetadata: Record<string, string>
}): unknown {
  const desc = JSON.parse(result.descriptor) as Record<string, unknown>
  const normDesc: Record<string, unknown> = { ...desc }
  normDesc.input_schema_object = desc.input_schema
    ? JSON.parse(desc.input_schema as string)
    : null
  normDesc.output_schema_object = desc.output_schema
    ? JSON.parse(desc.output_schema as string)
    : null
  delete normDesc.input_schema
  delete normDesc.output_schema
  return {
    tool_id: result.toolId,
    mcp_name: result.mcpName,
    bridge_metadata: result.bridgeMetadata,
    descriptor: normDesc,
  }
}

// napi-rs exports are `undefined` when the backing `mcp` Cargo feature isn't
// compiled. Skip (not fail opaquely with "not a function") when that's the
// case, so a slimmed build variant reads as skipped rather than broken.
const mcpBindingAvailable =
  typeof classifyMcpServer === 'function' && typeof lowerMcpTool === 'function'

describe.skipIf(!mcpBindingAvailable)('MCP helper golden vectors', () => {
  for (const c of FIXTURE.classify) {
    it(`classify: ${c.name}`, () => {
      const envs = Object.entries(c.envs).map(([key, value]) => ({ key, value }))
      const got = classifyMcpServer(
        c.program,
        c.args,
        envs,
        c.credential_override ?? undefined,
        c.force,
      )
      expect(got).toBe(c.expected_status)
    })
  }

  for (const c of FIXTURE.lower) {
    it(`lower: ${c.name}`, () => {
      const lowered = lowerMcpTool(
        JSON.stringify(c.tool),
        c.server_version,
        c.credential_status,
        c.substitutability,
      )
      expect(normalize(lowered)).toEqual(c.expected)
    })
  }
})
